# PaintSHOP UI logic
# Elliot Hershberg
# July 15, 2019

library(shiny)
library(shinythemes)
library(shinyjs)
library(magrittr)
library(shinycssloaders)
library(shinyBS)

# Define UI for application
shinyUI(fluidPage(
    
    theme = shinytheme("flatly"),
    useShinyjs(),
    
    # add CSS to the header to make horizontal rules visible
    tags$head(
      tags$style(HTML("hr {border-top: 3px solid #000000}"))
    ),
    
    navbarPage("PaintSHOP",
      tabPanel("About",
        div(tags$img(height = 300,
                 width = 300,
                 src = 'PaintSHOP-logo.png'),
            style="text-align: center;"),
        tags$h1("Welcome to PaintSHOP!"),
        tags$hr(),
        tags$h4("PaintSHOP is an application that enables the interactive design of Oligopaint probe libraries 
               for any experiment. The design process is broken down into a few simple steps:"),
        tags$br(),
        tags$ol(
          tags$li(tags$h4("Quickly retrieve the probes that cover your target(s) of interest")),
          tags$li(tags$h4("Interactively experiment with parameters that determine the specificity 
                          and coverage of your probe set. (optional)")),
          tags$li(tags$h4("Append the additional sequences to your probes that your experiment requires. (optional)")),
          tags$li(tags$h4("Download your newly created order file!"))
        ),
        tags$br(),
        tags$h4("The intended app workflow is shown below. If you want to know more about a specific feature in the app, 
                make sure to hover over it with your mouse to see an additional explanation. Happy FISHing!"),
        tags$br(),
        tags$img(height = 300, width = 1400,
                 src = 'workflow-diagram.png'),
        tags$h4("If you use PaintSHOP in your research, please cite our manuscript:")
      ),
      tabPanel("RNA Probe Design",
        sidebarLayout(
          sidebarPanel(
            selectInput("probeset", "Choose Probe Set:",
                         c("hg38 newBalance" = "probes/refseq/hg38_refseq_newBalance.tsv",
                           "hg19 newBalance" = "probes/refseq/hg19refseq_newBalance.tsv",
                           "mm10 newBalance" = "probes/refseq/mm10_refseq_newBalance.tsv",
                           "mm9 newBalance" = "probes/refseq/mm9_refseq_newBalance.tsv",
                           "dm6 newBalance" = "probes/refseq/dm6_refseq_newBalance.tsv",
                           "ce11 newBalance" = "probes/refseq/ce11_refseq_newBalance.tsv",
                           "OligoMiner hg38 balance" = "probes/refseq/hg38b_refseq.tsv",
                           "OligoMiner hg19 balance" = "probes/refseq/hg19b_refseq.tsv",
                           "2012 Oligopaints hg19" = "probes/refseq/hg19_2012_refseq.tsv",
                           "iFISH4U 40-mer hg19" = "probes/refseq/iFISH4U_refseq.tsv")),
            textInput("refseq_manual", label = h3("Enter RefSeq IDs manually"), 
                      value = "NM_020309, NM_015267, NM_018008, ..."),
            tags$h3("OR"),
            fileInput("refseq_file", label = h3("Upload a File With RefSeq IDs"), 
                      multiple = FALSE, accept = NULL,
                      width = NULL, buttonLabel = "Browse...",
                      placeholder = "No file selected"),
            radioButtons("refseq_manual_or_file", "Input type:",
                         c("Manual" = TRUE, "File" = FALSE),
                         selected = TRUE),
            actionButton("refseq_submit", label = tags$h3("Submit"),
                         icon("paper-plane"), width = "70%"),
            tags$br(),
            tags$br(),
            tags$p("Please upload a text file with one RefSeq accession per line.",
                   "For example:",
                   tags$br(),
                   tags$br(),
                   "NM_020309",
                   tags$br(),
                   "NM_015267",
                   tags$br(),
                   "NM_018008",
                   tags$br(),
                   "NM_000818",
                   tags$br(),
                   "..."),
            tags$hr(),
            tags$h3("Advanced Probe Settings (optional)"),
            radioButtons("repeat_seq", "Repeat:",
                         c("allowed" = TRUE, "none" = FALSE),
                         selected = FALSE),
            sliderInput("off_target", "Off-Target Score:",
                        min = 0, max = 1000,
                        value = 200),
            sliderInput("max_kmer", "Max K-mer Count:",
                        min = 0, max = 255,
                        value = 5),
            actionButton("restore_default", label = "Restore Default Parameters"),
            tags$hr(),
            tags$h3("Balance Set (optional)"),
            radioButtons("balance_show", "",
                         c("Show" = TRUE, "Hide" = FALSE),
                         selected = FALSE),
            hidden(sliderInput("balance_goal", "Number of probes per target:",
                        min = 15, max = 150,
                        value = 30)),
            hidden(radioButtons("balance_set", "Enable:",
                         c("Trim" = 1, "Balance" = 2, "Off" = 3),
                         selected = 3))
          ),
          mainPanel(
            plotOutput("count_plot") %>% withSpinner(color = "grey"),
            DT::dataTableOutput("intersect_table")
            
          )
        )
      ),
      tabPanel("DNA Probe Design",
        sidebarLayout(
         sidebarPanel(
           # change button values to path to complete newBalance probe set, not intersect
           selectInput("probeset_coord", "Choose Probe Set:",
                       c("hg38 newBalance" = "probes/all/hg38_all_newBalance",
                         "hg19 newBalance" = "probes/all/hg19_all_newBalance",
                         "mm10 newBalance" = "probes/all/mm10_all_newBalance",
                         "mm9 newBalance" = "probes/all/mm9_all_newBalance",
                         "dm6 newBalance" = "probes/all/dm6_all_newBalance",
                         "ce11 newBalance" = "probes/all/ce11_all_newBalance",
                         "OligoMiner hg38 balance" = "probes/all/hg38b_all",
                         "OligoMiner hg19 balance" = "probes/all/hg19b_all",
                         "2012 Oligopaints hg19" = "probes/all/hg19_2012_all",
                         "iFISH4U 40-mer hg19" = "probes/all/iFISH4U_all")),
           textInput("coord_manual", label = h3("Enter genomic coordinates manually "), 
                     value = "chr10:26216307-26304562, chr12:111034165-111350554, chr3:62369681-62373550"),
           tags$p("(Please enter chr:start-stop for each target. Optionally, you can specify which strand",
                  " to return, by entering chr:start-stop:+ or chr:start-stop:- for a target.)"),
           tags$h3("OR"),
           fileInput("coord_file", label = h3("Upload a file with genomic coordinates (BED-like format)"), 
                     multiple = FALSE, accept = NULL,
                     width = NULL, buttonLabel = "Browse...",
                     placeholder = "No file selected"),
           radioButtons("coord_manual_or_file", "Input type:",
                        c("Manual" = TRUE, "File" = FALSE),
                        selected = TRUE),
           actionButton("coord_submit", label = tags$h3("Submit"),
                        icon("paper-plane"), width = "70%"),
           tags$br(),
           tags$br(),
           tags$p("Please upload a BED-like file with one genomic coordinate per line.",
                  " The strand column is optional.",
                  "For example:", tags$br()),
           pre(includeText("example.bed")),
           tags$hr(),
           tags$h3("Advanced Probe Settings (optional)"),
           radioButtons("repeat_seq_coord", "Repeat:",
                        c("allowed" = TRUE, "none" = FALSE),
                        selected = FALSE),
           sliderInput("off_target_coord", "Off-Target Score:",
                       min = 0, max = 1000,
                       value = 200),
           sliderInput("max_kmer_coord", "Max K-mer Count:",
                       min = 0, max = 255,
                       value = 5),
           actionButton("coord_restore_default", label = "Restore Default Parameters"),
           tags$hr(),
           tags$h3("Balance Set (optional)"),
           radioButtons("coord_balance_show", "",
                        c("Show" = TRUE, "Hide" = FALSE),
                        selected = FALSE),
           hidden(sliderInput("coord_balance_goal", "Number of probes per target:",
                       min = 15, max = 10000,
                       value = 100)),
           hidden(radioButtons("coord_balance_set", "Enable:",
                        c("Trim" = 1, "Balance" = 2, "Off" = 3),
                        selected = 3))
         ),
         mainPanel(
           plotOutput("coord_count_plot") %>% withSpinner(color = "grey"),
           DT::dataTableOutput("coord_intersect_table")
           
         )
        )
      ),
      tabPanel("Append Sequences",
        sidebarLayout(
         sidebarPanel(
           # determine which design tab was used
           radioButtons("design_scheme", "Design Scheme:",
                        c("RNA Probe Design" = TRUE, "DNA Probe Design" = FALSE),
                        selected = TRUE),
           tags$hr(),
           ###########################################################################
           radioButtons("fpo_choice", label = h3("5' Outer Primer Sequence (O)"),
                        c("Append" = TRUE, "None" = FALSE),
                        selected = FALSE),
           hidden(radioButtons("fpo_orientation", "Orientation:",
                               c("Forward" = FALSE, 
                                 "Reverse complement" = TRUE),
                               selected = FALSE)),
           hidden(radioButtons("fpo_append_scheme", "Format:",
                               c("Same for all probes" = 1, 
                                 "Unique for each target" = 2,
                                 "Multiple per target" = 3,
                                 "Custom ranges" = 4),
                               selected = 1)),
           hidden(textInput("fpo_custom_ranges", label = "Custom Ranges (optional)", 
                            value = "1-100, 101-200, 201-300, ...")),
           hidden(sliderInput("fpo_n_per_target", "Number Per Target (optional)",
                              min = 1, max = 20, value = 1)),
           hidden(selectInput("fpo_sequence_select", label = "Select Sequence Set", 
                              choices = list("PaintSHOP 5' Outer Primer Set" = "appending/ps_of.tsv",
                                             "MERFISH Primer Set" = "appending/merfish_primers.tsv",
                                             "Custom Set" = FALSE), 
                              selected = "appending/ps_of.tsv")),
           hidden(fileInput("fpo_custom_file", label = "Upload Custom Set (optional)", 
                            multiple = FALSE, accept = NULL,
                            width = NULL, buttonLabel = "Browse...",
                            placeholder = "No file selected")),
           tags$hr(),
           ###########################################################################
           radioButtons("fpb_choice", label = h3("5' Bridge Sequence (B)"),
                        c("Append" = TRUE, "None" = FALSE),
                        selected = FALSE),
           hidden(radioButtons("fpb_orientation", "Orientation:",
                               c("Forward" = FALSE, 
                                 "Reverse complement" = TRUE),
                               selected = FALSE)),
           hidden(radioButtons("fpb_append_scheme", "Format:",
                               c("Same for all probes" = 1, 
                                 "Unique for each target" = 2,
                                 "Multiple per target" = 3,
                                 "Custom ranges" = 4),
                               selected = 1)),
           hidden(textInput("fpb_custom_ranges", label = "Custom Ranges (optional)",
                            value = "1-100, 101-200, 201-300, ...")),
           hidden(sliderInput("fpb_n_per_target", "Number Per Target (optional)",
                              min = 1, max = 20, value = 1)),
           hidden(selectInput("fpb_sequence_select", label = "Select Sequence Set", 
                              choices = list("PaintSHOP 5' Bridge Set" = "appending/ps_fpb.tsv",
                                             "PaintSHOP Full Bridge Set" = "appending/ps_bridges.tsv",
                                             "MERFISH Bridge Set" = "appending/merfish_bridges.tsv",
                                             "Mateo et al. 2019 Bridges" = "appending/Mateo2019_bridges.tsv",
                                             "Xia et al. 2019 Bridges" = "appending/Xia2019_bridges.tsv",
                                             "Custom Set" = FALSE), 
                              selected = "appending/ps_fpb.tsv")),
           hidden(fileInput("fpb_custom_file", label = "Upload Custom Set (optional)", 
                            multiple = FALSE, accept = NULL,
                            width = NULL, buttonLabel = "Browse...",
                            placeholder = "No file selected")),
           tags$hr(),
           ###########################################################################
           radioButtons("fpi_choice", label = h3("5' Inner Primer Sequence (I)"),
                        c("Append" = TRUE, "None" = FALSE),
                        selected = FALSE),
           hidden(radioButtons("fpi_orientation", "Orientation:",
                               c("Forward" = FALSE, 
                                 "Reverse complement" = TRUE),
                               selected = FALSE)),
           hidden(radioButtons("fpi_append_scheme", "Format:",
                               c("Same for all probes" = 1, 
                                 "Unique for each target" = 2,
                                 "Multiple per target" = 3,
                                 "Custom ranges" = 4),
                               selected = 1)),
           hidden(textInput("fpi_custom_ranges", label = "Custom Ranges (optional)",
                            value = "1-100, 101-200, 201-300, ...")),
           hidden(sliderInput("fpi_n_per_target", "Number Per Target (optional)",
                              min = 1, max = 20, value = 1)),
           hidden(selectInput("fpi_sequence_select", label = "Select Sequence Set", 
                              choices = list("PaintSHOP 5' Inner Primer Set" = "appending/ps_if.tsv",
                                             "MERFISH Primer Set" = "appending/merfish_primers.tsv",
                                             "Custom Set" = FALSE), 
                              selected = "appending/ps_if.tsv")),
           hidden(fileInput("fpi_custom_file", label = "Upload Custom Set (optional)", 
                            multiple = FALSE, accept = NULL,
                            width = NULL, buttonLabel = "Browse...",
                            placeholder = "No file selected")),
           tags$hr(),
           ###########################################################################
           
           # determine 3' encoding scheme
           radioButtons("tp_appending_choice", label = h3("3' Appending Scheme"),
                        c("SABER" = 1, "Primer/Bridge/Universal" = 2,
                          "None" = 3),
                        selected = 3),
           tags$hr(),
           ###########################################################################
           
           # SABER Sequence options
           hidden(radioButtons("saber_x", "Number of copies of concatemer:",
                               c("1x" = 1, "2x" = 2),
                               selected = 1)),
           hidden(radioButtons("saber_append_scheme", "Format:",
                               c("Same for all probes" = 1, 
                                 "Unique for each target" = 2,
                                 "Multiple per target" = 3,
                                 "Custom ranges" = 4),
                               selected = 1)),
           hidden(textInput("saber_custom_ranges", label = "Custom Ranges (optional)",
                            value = "1-100, 101-200, 201-300, ...")),
           hidden(sliderInput("saber_n_per_target", "Number Per Target (optional)",
                              min = 1, max = 20, value = 1)),

           ###########################################################################
           hidden(radioButtons("tpi_choice", label = h3("3' Inner Primer Sequence (I)"),
                        c("Append" = TRUE, "None" = FALSE),
                        selected = FALSE)),
           hidden(radioButtons("tpi_orientation", "Orientation:",
                               c("Forward" = FALSE, 
                                 "Reverse complement" = TRUE),
                               selected = TRUE)),
           hidden(radioButtons("tpi_append_scheme", "Format:",
                               c("Same for all probes" = 1, 
                                 "Unique for each target" = 2,
                                 "Multiple per target" = 3,
                                 "Custom ranges" = 4),
                               selected = 1)),
           hidden(textInput("tpi_custom_ranges", label = "Custom Ranges (optional)",
                            value = "1-100, 101-200, 201-300, ...")),
           hidden(sliderInput("tpi_n_per_target", "Number Per Target (optional)",
                              min = 1, max = 20, value = 1)),
           hidden(selectInput("tpi_sequence_select", label = "Select Sequence Set", 
                              choices = list("PaintSHOP 3' Inner Primer Set" = "appending/ps_ir.tsv",
                                             "MERFISH Primer Set" = "appending/merfish_primers.tsv",
                                             "Custom Set" = FALSE), 
                              selected = "appending/ps_ir.tsv")),
           hidden(fileInput("tpi_custom_file", label = "Upload Custom Set (optional)", 
                            multiple = FALSE, accept = NULL,
                            width = NULL, buttonLabel = "Browse...",
                            placeholder = "No file selected")),
           hidden(tags$hr(id = "tpi_choice_hr")),
           ###########################################################################
           hidden(radioButtons("tpb_choice", label = h3("3' Bridge Sequence (B)"),
                        c("Append" = TRUE, "None" = FALSE),
                        selected = FALSE)),
           hidden(radioButtons("tpb_orientation", "Orientation:",
                               c("Forward" = FALSE, 
                                 "Reverse complement" = TRUE),
                               selected = TRUE)),
           hidden(radioButtons("tpb_append_scheme", "Format:",
                               c("Same for all probes" = 1, 
                                 "Unique for each target" = 2,
                                 "Multiple per target" = 3,
                                 "Custom ranges" = 4),
                               selected = 1)),
           hidden(textInput("tpb_custom_ranges", label = "Custom Ranges (optional)",
                            value = "1-100, 101-200, 201-300, ...")),
           hidden(sliderInput("tpb_n_per_target", "Number Per Target (optional)",
                              min = 1, max = 20, value = 1)),
           hidden(selectInput("tpb_sequence_select", label = "Select Sequence Set", 
                              choices = list("PaintSHOP 3' Bridge Set" = "appending/ps_tpb.tsv",
                                             "PaintSHOP Bridge Set" = "appending/ps_bridges.tsv",
                                             "MERFISH Bridge Set" = "appending/merfish_bridges.tsv",
                                             "Mateo et al. 2019 Bridges" = "appending/Mateo2019_bridges.tsv",
                                             "Xia et al. 2019 Bridges" = "appending/Xia2019_bridges.tsv",
                                             "Custom Set" = FALSE), 
                              selected = "appending/ps_tpb.tsv")),
           hidden(fileInput("tpb_custom_file", label = "Upload Custom Set (optional)", 
                            multiple = FALSE, accept = NULL,
                            width = NULL, buttonLabel = "Browse...",
                            placeholder = "No file selected")),
           hidden(tags$hr(id = "tpb_choice_hr")),
           ###########################################################################
           hidden(radioButtons("tpo_choice", label = h3("3' Outer Primer Sequence (O)"),
                        c("Append" = TRUE, "None" = FALSE),
                        selected = FALSE)),
           hidden(radioButtons("tpo_orientation", "Orientation:",
                               c("Forward" = FALSE, 
                                 "Reverse complement" = TRUE),
                               selected = TRUE)),
           hidden(radioButtons("tpo_append_scheme", "Format:",
                               c("Same for all probes" = 1, 
                                 "Unique for each target" = 2,
                                 "Multiple per target" = 3,
                                 "Custom ranges" = 4),
                               selected = 1)),
           hidden(textInput("tpo_custom_ranges", label = "Custom Ranges (optional)", 
                            value = "1-100, 101-200, 201-300, ...")),
           hidden(sliderInput("tpo_n_per_target", "Number Per Target (optional)",
                              min = 1, max = 20, value = 1)),
           hidden(selectInput("tpo_sequence_select", label = "Select Sequence Set", 
                              choices = list("PaintSHOP 3' Outer Primer Set" = "appending/ps_or.tsv",
                                             "MERFISH Primer Set" = "appending/merfish_primers.tsv",
                                             "Custom Set" = FALSE), 
                              selected = "appending/ps_or.tsv")),
           hidden(fileInput("tpo_custom_file", label = "Upload Custom Set (optional)", 
                            multiple = FALSE, accept = NULL,
                            width = NULL, buttonLabel = "Browse...",
                            placeholder = "No file selected")),
           hidden(tags$hr(id = "tpo_choice_hr")),
           ###########################################################################
           actionButton("append_submit", label = "Append")
         ),
         mainPanel(
           # main panel for appending
           tags$img(height = 200,
                    width = 1000,
                    src = "probe-all.png"),
           DT::dataTableOutput("append_table")
         )
        )
      ),
      tabPanel("Download",
        sidebarLayout(
          sidebarPanel(
            radioButtons("appended_tf", "Appended Sequences:",
                         c("Yes" = TRUE, "No" = FALSE),
                         selected = FALSE),
            # determine which type of design should be downloaded
            radioButtons("download_design_scheme", "Design Scheme:",
                         c("RNA Probe Design" = TRUE, "DNA Probe Design" = FALSE),
                         selected = TRUE),
            selectInput("download_choice", "Choose a file to download:",
                        choices = c("None Selected" = 1,
                                    "Order File" = 2,
                                    "Appending File" = 3,
                                    "Full Probe File" = 4,
                                    "Citation File" = 5),
                        selected = 1),
            downloadButton("download_file", "Download")
          ),
          mainPanel(
            DT::dataTableOutput("download_table")
          )
        )
      ),
      tabPanel("Resources Available",
               "As part of our effort to provide open resources for the oligo-FISH community, ",
               "all probe sets and appending sets used in the backend of PaintSHOP are available ",
               "for download here: ", 
               tags$a("link", href = "https://github.com/elliothershberg/PaintSHOP_resources",
                      target = "_blank"),
               tags$br(),
               tags$br(),
               "Our goal in designing PaintSHOP was to make easy things easy, and hard things possible. ",
               "A key use case of the resources we have made available is to download the sets of sequences ",
               "that can be appended, create custom appending files, and use those files in PaintSHOP. ",
               "With this flexibility, practically all appending design schemes should be possible."
      ),
      tabPanel("Documentation",
        tags$h3("Click on a topic to expand the description", align = "center"),
        tags$br(),
        bsCollapse(id = "tutorial_collapse", multiple = TRUE,
          bsCollapsePanel("RNA Probe Design", 
                         "The RNA Probe Design tab allows you to enter RefSeq annotations ",
                         "in order to retrieve the probes for your target or list of targets. ",
                         "You can either enter the annotations manually into the text box provided, ",
                         "or upload a file with one annotation per line. If entering annotations ",
                         "manually, please separate each entry with a comma and a space, as shown ",
                         "in the example entry. Once you have entered your annotations, select which ",
                         " input option you used, and press the Submit button to retrieve your probes. ",
                         tags$br(),
                         tags$br(),
                         tags$img(width = 500,
                                  height = 150,
                                  src = "annotation-input.png"),
                         tags$br(),
                         tags$br(),
                         "PaintSHOP will return a density plot of probe coverage, and a dynamic table ",
                         "with your probes. The density plot is intended to provide a visual estimate ",
                         "of how well covered your annotations are.",
                         tags$br(),
                         tags$br(),
                         "The table has the following columns: ",
                         tags$hr(),
                         tags$b("refseq:"), 
                         " the RefSeq ID the probe targets",
                         tags$br(),
                         tags$b("chrom:"), 
                         " the chromosome of the probe sequence",
                         tags$br(),
                         tags$b("start:"), 
                         " the start coordinate of the probe sequence",
                         tags$br(),
                         tags$b("stop:"), 
                         " the stop coordinate of the probe sequence",
                         tags$br(),
                         tags$b("sequence:"), 
                         " The DNA sequence of the oligo probe",
                         tags$br(),
                         tags$b("Tm:"), 
                         " The melting temperature of the probe sequence",
                         tags$br(),
                         tags$b("on_target:"), 
                         " The on-target score generated by the Homology Optimization Pipeline",
                         tags$br(),
                         tags$b("off_target:"), 
                         " The off-target score generated by the Homology Optimization Pipeline",
                         tags$br(),
                         tags$b("repeat_seq:"), 
                         " Whether or not the probe sequence contains bases flagged as repetitive by",
                         " RepeatMasker. 0 = False, 1 = True",
                         tags$br(),
                         tags$b("max_kmer:"), 
                         " The maximum number of occurences of all 18-mers in the probe sequence",
                         "in the genome it targets",
                         tags$br(),
                         tags$b("probe_strand:"), 
                         " The strand orientation of the probe sequence. Positive (+) or negative (-)",
                         tags$hr(),
                         tags$b("Note:"), 
                         " If your target is on the + strand, PaintSHOP will automatically return ",
                         "the - strand probe sequence. This is to ensure your FISH experiment works ",
                         "the way you want!",
                         tags$br(),
                         tags$br(),
                         "Also note that the table can be searched, resized, and paged through. ",
                         tags$br(),
                         tags$br(),
                         "For more information about the advanced settings and set balancing features, ",
                         "Please read those topic descriptions.",
                         tags$hr(),
                         tags$h3("Probe Sets:"),
                         tags$br(),
                         tags$b("newBalance (RNA):"), "probe sets for the hg38, hg19, mm10, mm9, dm6 and ce11 genomes",
                         "with a length window of 30-37 nucleotides and a Tm of window of 42-47 degrees. These parameters ",
                         "were selected to optimize probe coverage and hybridization. ",
                         "For more information on these new probe sets, please refer to the PaintSHOP manuscript.",
                         tags$br(),
                         tags$br(),
                         tags$b("OligoMiner (RNA):"), " the 'balance' probe sets generated by OligoMiner ",
                         "for the hg39 and hg19 reference genomes. These probes have a length window of ",
                         "35-41 nucleotides, and a Tm window of 42-47 degrees. ",
                         "For more information on these probes, please refer to the OligoMiner manuscript.",
                         tags$br(),
                         tags$br(),
                         tags$b("2012 Oligopaints (RNA):"), " the original Oligopaints genome-scale ",
                         "probe set from the Beliveau et al. 2012 PNAS publication. The probes have a ",
                         "length of 32 bases, and have an approximate Tm window of 34-42 degrees. ",
                         "For more information on this probe set, please refer to the 2012 Oligopaints ",
                         "publication.",
                         tags$br(),
                         tags$br(),
                         tags$b("iFISH4U (RNA):"), " the Full 40-mer probe set from ",
                         tags$a("iFISH4U", href = "http://ifish4u.org/", target = "_blank"), ".",
                         tags$br(),
                         tags$br(),
                         "All probe sets in the RNA Probe Design tab have been intersected with the ",
                         "RefSeq for their respective reference genome in order to provide quick ",
                         "retrieval of probes for annotations. To retrieve probes target regions ",
                         "outside of these annotations, please use the DNA Probe Design tab.",
                         
                         style = "primary"),
          bsCollapsePanel("DNA Probe Design", 
                         "The RNA Probe Design tab allows you to enter genomic coordinates ",
                         "in order to retrieve the probes for your target or list of targets. ",
                         "You can either enter the coordinates manually into the text box provided, ",
                         "or upload a file with the format shown in the example. If entering coordinates ",
                         "manually, please separate each entry with a comma and a space, as shown ",
                         "in the example entry. Once you have entered your annotations, select which ",
                         " input option you used, and press the Submit button to retrieve your probes. ",
                         tags$br(),
                         tags$br(),
                         tags$img(width = 500,
                                  height = 150,
                                  src = "annotation-input.png"),
                         tags$br(),
                         tags$br(),
                         "PaintSHOP will return a density plot of probe coverage, and a dynamic table ",
                         "with your probes. The density plot is intended to provide a visual estimate ",
                         "of how well covered your targets are.",
                         tags$br(),
                         tags$br(),
                         "The table has the following columns: ",
                         tags$hr(),
                         tags$b("chrom:"), 
                         " the chromosome of the probe sequence",
                         tags$br(),
                         tags$b("start:"), 
                         " the start coordinate of the probe sequence",
                         tags$br(),
                         tags$b("stop:"), 
                         " the stop coordinate of the probe sequence",
                         tags$br(),
                         tags$b("sequence:"), 
                         " The DNA sequence of the oligo probe",
                         tags$br(),
                         tags$b("Tm:"), 
                         " The melting temperature of the probe sequence",
                         tags$br(),
                         tags$b("on_target:"), 
                         " The on-target score generated by the Homology Optimization Pipeline",
                         tags$br(),
                         tags$b("off_target:"), 
                         " The off-target score generated by the Homology Optimization Pipeline",
                         tags$br(),
                         tags$b("repeat_seq:"), 
                         " Whether or not the probe sequence contains bases flagged as repetitive by",
                         " RepeatMasker. 0 = False, 1 = True",
                         tags$br(),
                         tags$b("max_kmer:"), 
                         " The maximum number of occurences of all 18-mers in the probe sequence",
                         "in the genome it targets",
                         tags$br(),
                         tags$b("probe_strand:"), 
                         " The strand orientation of the probe sequence. Positive (+) or negative (-)",
                         tags$hr(),
                         tags$b("Note:"), 
                         " If you specify probe strand (+ or -) in either the manual or file entry, ",
                         "PaintSHOP will return the probe in the same orientation as what you enter. ",
                         "For example specifying + will return +. You are specifying the probe strand, ",
                         "not the target strand.",
                         tags$br(),
                         tags$br(),
                         "Also note that the table can be searched, resized, and paged through. ",
                         tags$br(),
                         tags$br(),
                         "For more information about the advanced settings and set balancing features, ",
                         "Please read those topic descriptions.",
                         tags$hr(),
                         tags$h3("Probe Sets:"),
                         tags$br(),
                         tags$b("newBalance (DNA):"), "probe sets for the hg38, hg19, mm10, mm9, dm6 and ce11 genomes",
                         "with a length window of 30-37 nucleotides and a Tm of window of 42-47 degrees. These parameters ",
                         "were selected to optimize probe coverage and hybridization. ",
                         "For more information on these new probe sets, please refer to the PaintSHOP manuscript.",
                         tags$br(),
                         tags$br(),
                         tags$b("OligoMiner (DNA):"), " the 'balance' probe sets generated by OligoMiner ",
                         "for the hg39 and hg19 reference genomes. These probes have a length window of ",
                         "35-41 nucleotides, and a Tm window of 42-47 degrees. ",
                         "For more information on these probes, please refer to the OligoMiner manuscript.",
                         tags$br(),
                         tags$br(),
                         tags$b("2012 Oligopaints (DNA):"), " the original Oligopaints genome-scale ",
                         "probe set from the Beliveau et al. 2012 PNAS publication. The probes have a ",
                         "length of 32 bases, and have an approximate Tm window of 34-42 degrees. ",
                         "For more information on this probe set, please refer to the 2012 Oligopaints ",
                         "publication.",
                         tags$br(),
                         tags$br(),
                         tags$b("iFISH4U (DNA):"), " the Full 40-mer probe set from ",
                         tags$a("iFISH4U", href = "http://ifish4u.org/", target = "_blank"), ".",
                         tags$br(),
                         tags$br(),
                         "All probe sets in the DNA Probe Design tab include all probes targeting ",
                         "the entire reference genome that they target. It takes longer to load ",
                         "and search these sets, but probes can be designed against any coordinates ",
                         "in a given genome. You can use the RNA Probe Design for faster retrieval ",
                         "of probes targeting specific RefSeq coordinates.",
                         
                         style = "primary"),
          bsCollapsePanel("Advanced Probe Settings",
                          "In some cases, you may find that you want to increase the number of probes ",
                          "returned for your targets. PaintSHOP provides a set of advanced features to ",
                          "enable this flexibility. In computational probe design there is an inherent ",
                          "trade-off between probe coverage and specifity. Probe specificity is the ",
                          "likelihood that a probe hybridizes to its intended target instead of at another ",
                          "site in the genome. A greater emphasis on probe specificity inevitably filters ",
                          "probes, reducing coverage. In order to have more control over this design choice, ",
                          "PaintSHOP provides three parameters: 1) repeat inclusion, 2) off-target score, and ",
                          "3) the maximum k-mer count. These three parameters are described below.",
                          tags$hr(),
                          tags$b("Repeat:"),
                          tags$a("RepeatMasker", href = "http://www.repeatmasker.org/", target = "_blank"),
                          " is a program that identifies the presence of repetitive elements in a given DNA ",
                          "sequence. The human genome has been annotated by RepeatMasker. Previous probe design ",
                          "tools have excluded repetitive sequences. PaintSHOP provides the option to allow for ",
                          "probes which contain bases that have been flagged as repetitive, if it is necessary ",
                          "to have enough probes for a target that is challenging to cover. By default, repeat mode ",
                          "is set to off.",
                          tags$br(),
                          tags$br(),
                          tags$b("Off-Target Score:"),
                          " One important component of PaintSHOP is the Homology Optimization Pipeline (HOP). ",
                          "The pipeline is used when creating new probe sets to create an on-target and off-target ",
                          "score for every probe identified. We have developed a machine learning model to approximate ",
                          "nucleic acid thermodynamics. For the on-target score, the model is used to score the ",
                          "likelihood (0-100) that a probe is likely to hybridize at its intended target. ",
                          "For the off-target score, we start by searching for up to 100 possible alignments ",
                          "for each candidate probe. Any candidate with greater than 100 possible alignments is ",
                          "discarded. Next, we use our model to generate a score for the likelihood of hybridization ",
                          "at each possible site. We sum these scores, generating an off-target score between 0 and ",
                          "10,000. By default, PaintSHOP sets the default maximum off-target score to 200. The ",
                          "off-target score slider can be used to make this value more or less stringent, depending ",
                          "on the experiment. The probe table and density plot will dynamically update, providing ",
                          "information on how the parameters chosen affect the probe set.",
                          tags$br(),
                          tags$br(),
                          tags$b("Max K-mer count:"),
                          " A k-mer is a substring of a DNA sequence of length k. We use ",
                          tags$a("JELLYFISH", href = "https://www.cbcb.umd.edu/software/jellyfish/", target = "_blank"),
                          " to count how many times each 18-mer substring in a given probe occurs. ",
                          "in the genome it targets. ",
                          "The maximum value of all 18-mer counts is another way to control probe specifity, ",
                          "And can identify problematic problematic substrings that other alignment approaches ",
                          "may miss. By default, PaintSHOP uses a maximum k-mer count of 5. This can be changed ",
                          " using the slider, and the probe table and density plot will dynamically update.",
                          tags$hr(),
                          tags$b("Note:"),
                          " At any time, the Restore Default Parameters button can be used to reset the default ",
                          "values if you want to remove changes you have made.",
                          
                          style = "primary"),
          bsCollapsePanel("Balancing a Probe Set",
                          "In some instances, you may want to even out your probe set, by selecting a certain ",
                          "number of probes per target. PaintSHOP offers two features for balancing a probe ",
                          "set: 1) trim, and 2) balance. For either option, use the slider to choose how many ",
                          "probes you want per target. The trim option simply ranks the probes for each target ",
                          "by off-target score, and keeps however many you need, removing the rest. The balance ",
                          "option will behave the same way for targets with enough probes, and will automatically ",
                          "relax stringency parameters for targets without enough probes with the currently ",
                          "selected parameters to return enough probes to meet the target, and trim the probes ",
                          "from targets with too many probes.", tags$b("Importantly, "), "this means that ",
                          "changes from the Advanced Probe Settings section will be overriden using the ",
                          "balance feature.",
                          
                          style = "primary"),
          bsCollapsePanel("Appending Sequences",
                          "PaintSHOP provides a suite of functionality for appending the necessary sequences ",
                          "to your probes to carry out your experiment. The following diagram shows where you ",
                          "can append sequences to your probes:",
                          tags$br(),
                          tags$br(),
                          tags$img(height = 200,
                                   width = 800,
                                   src = "probe-all.png"),
                          tags$br(),
                          tags$br(),
                          "The probe sequence itself is the Homology Region (H) in the center. If you are ",
                          "going to amplify your probes from an oligo pool using PCR, you can add forward ",
                          "primers to the 5' I location, and reverse primers to the 3' I location ",
                          "(I stands for Inner Primer). The next regions where sequences can be added ",
                          "are the bridge regions (B) on both the 5' and 3' sides of the probe. ",
                          "Other names for this include read out, ear, and barcode. This sequence can be used ",
                          "as a location for secondary oligos to bind to. Sequences can also be appended to ",
                          "the outer region (O) on the 5' and 3' sides. One useful application of this is ",
                          "to include region/target specific primer pairs to be able to amplify only specific ",
                          "portions of your oligo pool at certain times in your experimental workflow. ",
                          "You can also choose to add ", tags$a("SABER", href = "http://saber.fish/", target = "_blank"),
                          " concatemer sequences to your probes. ",
                          tags$br(),
                          tags$br(),
                          tags$b("Every appending location is optional."),
                          tags$br(),
                          tags$br(),
                          "To append sequences, first select whether you used the RNA Probe Design or DNA ",
                          " Probe Design tab. ",
                          "If you want to append sequences to a specific region, switch the radio button for ",
                          "the region from None to Append. Once you choose to append a sequence, a set of ",
                          "options will appear. For example, once the 5' Outer Primer Sequence is selected, ",
                          "the appending menu will look like this: ",
                          tags$br(),
                          tags$br(),
                          tags$img(height = "25%",
                                   width = "25%",
                                   src = "appending-documentation.png"),
                          tags$br(),
                          tags$br(),
                          "The ",
                          tags$b("Orientation "),
                          "option lets you choose between appending the forward (5' to 3') ",
                          "orientation of the sequence, and the reverse complement of the sequence. The ",
                          "default orientation is forward for the 5' regions and reverse complement for ",
                          "the 3' regions. SABER sequences have a specific orientation and don't have ",
                          "this option.",
                          tags$br(),
                          tags$br(),
                          "The ", tags$b("Format "), "orientation is very important. This determines ",
                          "how sequences from whatever set you choose will be used. If you choose ",
                          "'Same for all probes', the first sequence in the selected sequence set ",
                          "will be appended to all probes for that region. An example of this use ",
                          "case would be appending the same forward and reverse primer pair to your ",
                          "entire probe set. The next option is 'Unique for each target'. If this ",
                          "option is selected, each target in your probe set will have a ",
                          "unique appended to it. For example, if you are appending a primer ",
                          "sequence to a probe set with 10 targets, the first target will have ",
                          "the first sequence in the selected set appended, while the second ",
                          "sequence will be appended to the second target, all the way up to the ",
                          "tenth target receiving the tenth sequence. This can be used to append ",
                          "a unique bridge sequence to each target, for example. ",
                          "'Multiple per target' allows you to add more than one sequence to each ",
                          "target. Assume you have set [A, B, C] of sequences to append, and your ",
                          "target has probes 1, 2, 3, 4, 5. You select to append 3 sequences to ",
                          "this target. This option will append A1, B2, C3, A4, B5. This option ",
                          "is intended to support oligo FISH technologies such as seqFISH and ",
                          "MERFISH. Choose how many probes per target with the ",
                          tags$b("Number Per Target"), " slider.",
                          tags$br(),
                          tags$br(),
                          "If none of the built-in appending formats are quite right for your ",
                          "experiment, you can use the ", tags$b("Custom Ranges "),
                          "option. First, select 'Custom ranges' as your ", tags$b("Format "),
                          "choice. The ranges that you enter in the text box correspond ",
                          "to ranges of your probe set that should get a unique probe. ",
                          "For example, for a probe set with 200 probes, you could add ",
                          "a unique sequence to each half by specifying '1-100, 101-200' ",
                          "as a custom range. You can use the probe table returned in the ",
                          "RNA/DNA Probe Design tab to design custom ranges.",
                          tags$br(),
                          tags$br(),
                          "To choose what set of sequences to append for a region, use the ",
                          tags$b("Select Sequence Set "), "drop-down menu. The menu will be ",
                          "populated with sets specific to the region that you chose to append ",
                          "sequences to. Descriptions of the different sequence sets can be ",
                          "found below. Alternatively, you can choose 'Custom Set' from the ",
                          "drop down and use the ", tags$b("Upload Custom Set "),
                          "option to use your own sequence file. Upload a text file with ",
                          "one sequence per line.",
                          tags$br(),
                          tags$br(),
                          "You can choose between appending Primer/Bridge/Universal sequence(s) ",
                          "or a SABER concatemer sequence to the 3' side of your probe. The ",
                          "Primer/Bridge/Universal sequences have the same options on both the ",
                          "5' and 3' side. The SABER interface is simpler, since there are less ",
                          "choices to make. The two options are the number of copies of the ",
                          "concatemer (1 or 2), and the ", tags$b("Format"), " that should be ",
                          "used to append the sequences.",
                          tags$br(),
                          tags$br(),
                          "Once you have selected the appending options necessary for your ",
                          "experiment, click the ", tags$b("Append"), " button at the bottom ",
                          "of the side panel. A dynamic table will appear, showing which ",
                          "sequences where appended. The first column of the table is a list of ",
                          "your targets. The other columns are determined by which sequences ",
                          "you append. Each entry in the table has the following format: ",
                          "set_type#", " where the sets are: ",
                          tags$br(),
                          tags$br(),
                          "ps = PaintSHOP",
                          tags$br(),
                          "saber_1x = SABER PER primers",
                          tags$br(),
                          "saber_2x = SABER PER primers with two copies of a given concatemer",
                          tags$br(),
                          "merfish = Sets from[MERFISH site",
                          tags$br(),
                          "Mateo2019 = Sets from Mateo et al. 2019",
                          tags$br(),
                          "Xia2019 = Sets from Xia et al. 2019",
                          tags$br(),
                          tags$br(),
                          "And the types are:",
                          tags$br(),
                          tags$br(),
                          "of = outer forward primer",
                          tags$br(),
                          "or = outer reverse primer",
                          tags$br(),
                          "if = inner forward primer",
                          tags$br(),
                          "ir = inner reverse primer",
                          tags$br(),
                          "fpb = 5' bridge sequence",
                          tags$br(),
                          "tpb = 3' bridge sequence",
                          tags$br(),
                          "primer = primer with no specified forward/reverse",
                          tags$br(),
                          "bridge = bridge set that can be used on either 5' or 3' side or used to make a custom set",
                          tags$br(),
                          tags$br(),
                          "For example: ps_of1 is the first outer forward primer of the PaintSHOP set. ",
                          "The actual sequence is also shown for each table entry as well.",
                          tags$hr(),
                          tags$h3("Appending Sets:"),
                          tags$br(),
                          tags$b("PaintSHOP 5' Outer Primer Set:"), " a set of 10 forward primers ",
                          "for the 5' O position. This set is paired with the PaintSHOP 3' Outer ",
                          "Primer Set. ",
                          tags$br(),
                          tags$br(),
                          tags$b("MERFISH Primer Set:"), " a set of 318 primer sequences from the ",
                          tags$a("MERFISH", href = "http://zhuang.harvard.edu/merfish.html", target = "_blank"),
                          " resources. This set is available for use at all primer locations. ",
                          tags$br(),
                          tags$br(),
                          tags$b("PaintSHOP Full Bridge Set:"), " a set of 1065 new bridge sequences ",
                          "provided with PaintSHOP that are designed to be orthogonal to all other ",
                          "sequences in the set. This set is available for use on both the 5' and 3' ",
                          " Bridge regions.",
                          tags$br(),
                          tags$br(),
                          tags$b("PaintSHOP 5' Bridge Set:"), " 533 of the PaintSHOP bridge sequences ",
                          "that can be used for the 5' Bridge region if you are going to append ",
                          "different bridge sequences to the 5' and 3' sides of your probe.",
                          tags$br(),
                          tags$br(),
                          tags$b("PaintSHOP 3' Bridge Set:"), " 532 of the PaintSHOP bridge sequences ",
                          "that can be used for the 3' Bridge region if you are going to append ",
                          "different bridge sequences to the 5' and 3' sides of your probe.",
                          tags$br(),
                          tags$br(),
                          tags$b("MERFISH Bridge Set:"), " a set of 16 bridge sequences from the ",
                          tags$a("MERFISH", href = "http://zhuang.harvard.edu/merfish.html", target = "_blank"),
                          " resources. This set is available for use on both the 5' and 3' ",
                          " Bridge regions.",
                          tags$br(),
                          tags$br(),
                          tags$b("Mateo et al. 2019 Bridges:"), " a set of 199 bridge sequences provided with the ",
                          tags$a("Mateo et al. 2019", href = "https://www.nature.com/articles/s41586-019-1035-4", 
                                 target = "_blank"),
                          " publication in Nature. These sequences are available for use on both the 5' and 3' ",
                          " Bridge regions.",
                          tags$br(),
                          tags$br(),
                          tags$b("Xia et al. 2019 Bridges:"), " a set of 70 bridge sequences provided with the ",
                          tags$a("Xia et al. 2019", href = "https://www.pnas.org/content/116/39/19490", target = "_blank"),
                          " publication in PNAS. These sequences are available for use on both the 5' and 3' ",
                          " Bridge regions.",
                          tags$br(),
                          tags$br(),
                          tags$b("PaintSHOP 5' Inner Primer Set:"), " a set of 74 forward primers ",
                          "for the 5' I position. This set is paired with the PaintSHOP 3' Inner ",
                          "Primer Set. ",
                          tags$br(),
                          tags$br(),
                          tags$b("PaintSHOP 3' Inner Primer Set:"), " a set of 74 reverse primers ",
                          "for the 3' I position. This set is paired with the PaintSHOP 5' Inner ",
                          "Primer Set. ",
                          tags$br(),
                          tags$br(),
                          tags$b("PaintSHOP 3' Outer Primer Set:"), " a set of 10 reverse primers ",
                          "for the 3' O position. This set is paired with the PaintSHOP 5' Outer ",
                          "Primer Set. ",
                          tags$br(),
                          tags$br(),
                          "It is important to note that PaintSHOP will start at the top of the ",
                          "list of whatever set you choose at each location. If you would like ",
                          "to use different primers from the same set at different locations, you may ",
                          "want to download the set from the ", tags$b("Resources"), " tab ",
                          "and create custom files with the sequences that you would like.",
                          tags$br(),
                          tags$br(),
                          "If you have sequences that you would like to be included in the PaintSHOP ",
                          "application, please reach out to us!",
                          
                          style = "primary"),
          bsCollapsePanel("Downloading Designs",
                          "Once you have finished designing your probe set and appending any sequences ",
                          "necessary for your experiment, you can visit the ", tags$b("Download"),
                          " tab. The first step when downloading is to select whether or not you ",
                          "appended sequences to your probe set, and whether you used the RNA ",
                          "or DNA Probe Design tab. ", " Next, you can choose the type of ",
                          "file you would like to download from the drop-down menu. PaintSHOP ",
                          "provides four different files for download:",
                          tags$br(),
                          tags$br(),
                          tags$b("1. Order File"),
                          tags$br(),
                          tags$b("2. Appending File"),
                          tags$br(),
                          tags$b("3. Full Probe File"),
                          tags$br(),
                          tags$b("4. Citation File"),
                          tags$br(),
                          tags$br(),
                          "The ", tags$b("Order File"), " is formatted to be ready to be sent directly ",
                          "to a company as an order for an oligo pool/library. The format is simple. ",
                          "The file has two columns. The first contains a unique identifier for each ",
                          "probe sequence, generated by concatenating the chromosome and the start ",
                          "coordinate of the probe. The second column is the actual sequence.",
                          tags$br(),
                          tags$br(),
                          "The ", tags$b("Appending File"), " is of the same format as the table returned ",
                          "in the appending tab. This can be useful to download as documentation of your ",
                          "final appending designs.",
                          tags$br(),
                          tags$br(),
                          "The ", tags$b("Full Probe File"), " is of the same format as whichever design ",
                          "tab you used to select your probes. This file can serve as a reference with ",
                          "full information about the probe sequences in your set.",
                          tags$br(),
                          tags$br(),
                          "The ", tags$b("Citation File"), " contains formatted citations and BibTeX entries ",
                          "for the publications associated with all resources used in the application. ",
                          "The BibTex entries can be imported directly into most reference managers.",
                          "Please cite the publications that are associated with the resources that you ",
                          "used. This recognizes the effort involved in making these resources freely ",
                          "available to accelerate research.",
                          
                          style = "primary"),
          bsCollapsePanel("Resources Available",
                          "As part of our effort to provide open resources for the oligo-FISH community, ",
                          "all probe sets and appending sets used in the backend of PaintSHOP are available ",
                          "for download here: ", 
                          tags$a("link", href = "https://github.com/elliothershberg/PaintSHOP_resources",
                                 target = "_blank"),
                          tags$br(),
                          tags$br(),
                          "Our goal in designing PaintSHOP was to make easy things easy, and hard things possible. ",
                          "A key use case of the resources we have made available is to download the sets of sequences ",
                          "that can be appended, create custom appending files, and use those files in PaintSHOP. ",
                          "With this flexibility, practically all appending design schemes should be possible.",
                          
                          style = "primary")

        )
        
      )
    )
))
