# PaintSHOP UI logic
# Elliot Hershberg
# July 15, 2019

library(shiny)
library(shinythemes)
library(shinyjs)
library(magrittr)
library(shinycssloaders)

# Define UI for application that draws a histogram
shinyUI(fluidPage(
    
    theme = shinytheme("flatly"),
    useShinyjs(),
    
    # add CSS to the header to make horizontal rules visible
    tags$head(
      tags$style(HTML("hr {border-top: 3px solid #000000}"))
    ),
    
    navbarPage("PaintSHOP",
      tabPanel("About",
        div(tags$img(height = 300,
                 width = 300,
                 src = 'PaintSHOP-logo.png'),
            style="text-align: center;"),
        tags$h1("Welcome to PaintSHOP!"),
        tags$hr(),
        tags$h4("PaintSHOP is an application that enables the interactive design of Oligopaint probe libraries 
               for any experiment. The design process is broken down into a few simple steps:"),
        tags$br(),
        tags$ol(
          tags$li(tags$h4("Quickly retrieve the probes that cover your target(s) of interest")),
          tags$li(tags$h4("Interactively experiment with parameters that determine the specificity 
                          and coverage of your probe set. (optional)")),
          tags$li(tags$h4("Append the additional sequences to your probes that your experiment requires. (optional)")),
          tags$li(tags$h4("Download your newly created order file!"))
        ),
        tags$br(),
        tags$h4("The intended app workflow is shown below. If you want to know more about a specific feature in the app, 
                make sure to hover over it with your mouse to see an additional explanation. Happy FISHing!"),
        tags$br(),
        tags$img(height = 300, width = 1400,
                 src = 'workflow-diagram.png')
      ),
      tabPanel("RNA Probe Design",
        sidebarLayout(
          sidebarPanel(
            selectInput("probeset", "Choose Probe Set:",
                         c("hg38 newBalance" = "probes/refseq/hg38_refseq_newBalance.tsv",
                           "hg19 newBalance" = "probes/refseq/hg19_refseq_newBalance.tsv",
                           "mm10 newBalance" = "probes/refseq/mm10_refseq_newBalance.tsv",
                           "mm9 newBalance" = "probes/refseq/mm9_refseq_newBalance.tsv",
                           "dm6 newBalance" = "probes/refseq/dm6_refseq_newBalance.tsv",
                           "ce11 newBalance" = "probes/refseq/ce11_refseq_newBalance.tsv",
                           "OligoMiner hg38 balance" = "probes/refseq/hg38b_refseq.tsv",
                           "OligoMiner hg19 balance" = "probes/refseq/hg19b_refseq.tsv",
                           "2012 Oligopaints hg19" = "probes/refseq/hg19_2012_refseq.tsv",
                           "iFISH4U 40-mer hg19" = "probes/refseq/iFISH4U_refseq.tsv")),
            textInput("refseq_manual", label = h3("Enter RefSeq IDs manually"), 
                      value = "NM_020309, NM_015267, NM_018008, ..."),
            tags$h3("OR"),
            fileInput("refseq_file", label = h3("Upload a File With RefSeq IDs"), 
                      multiple = FALSE, accept = NULL,
                      width = NULL, buttonLabel = "Browse...",
                      placeholder = "No file selected"),
            radioButtons("refseq_manual_or_file", "Input type:",
                         c("Manual" = TRUE, "File" = FALSE),
                         selected = FALSE),
            actionButton("refseq_submit", label = "Submit"),
            tags$p("Please upload a text file with one RefSeq accession per line.",
                   "For example:",
                   tags$br(),
                   tags$br(),
                   "NM_020309",
                   tags$br(),
                   "NM_015267",
                   tags$br(),
                   "NM_018008",
                   tags$br(),
                   "NM_000818",
                   tags$br(),
                   "..."),
            tags$hr(),
            tags$h3("Advanced Probe Settings (optional)"),
            radioButtons("repeat_seq", "Repeat:",
                         c("allowed" = TRUE, "none" = FALSE),
                         selected = FALSE),
            sliderInput("off_target", "Off-Target Score:",
                        min = 0, max = 1000,
                        value = 200),
            sliderInput("max_kmer", "Max K-mer Count:",
                        min = 0, max = 255,
                        value = 5),
            actionButton("restore_default", label = "Restore Default Parameters"),
            tags$hr(),
            tags$h3("Balance Set (optional)"),
            radioButtons("balance_show", "",
                         c("Show" = TRUE, "Hide" = FALSE),
                         selected = FALSE),
            hidden(sliderInput("balance_goal", "Number of probes per target:",
                        min = 15, max = 60,
                        value = 30)),
            hidden(radioButtons("balance_set", "Enable:",
                         c("Trim" = 1, "Balance" = 2, "Off" = 3),
                         selected = 3))
          ),
          mainPanel(
            plotOutput("count_plot") %>% withSpinner(color = "grey"),
            DT::dataTableOutput("intersect_table")
            
          )
        )
      ),
      tabPanel("DNA Probe Design",
        sidebarLayout(
         sidebarPanel(
           # change button values to path to complete newBalance probe set, not intersect
           selectInput("probeset_coord", "Choose Probe Set:",
                       c("hg38 newBalance" = "probes/all/hg38_all_newBalance.tsv",
                         "hg19 newBalance" = "probes/all/hg19_all_newBalance.tsv",
                         "mm10 newBalance" = "probes/all/mm10_all_newBalance.tsv",
                         "mm9 newBalance" = "probes/all/mm9_all_newBalance.tsv",
                         "dm6 newBalance" = "probes/all/dm6_all_newBalance.tsv",
                         "ce11 newBalance" = "probes/all/ce11_all_newBalance.tsv",
                         "OligoMiner hg38 balance" = "probes/all/hg38b_all.tsv",
                         "OligoMiner hg19 balance" = "probes/all/hg19b_all.tsv",
                         "2012 Oligopaints hg19" = "probes/all/hg19_2012_all.tsv",
                         "iFISH4U 40-mer hg19" = "probes/all/iFISH4U_all.tsv")),
           textInput("coord_manual", label = h3("Enter genomic coordinates manually "), 
                     value = "chr10:26216307-26304562, chr12:111034165-111350554, chr3:62369681-62373550, ..."),
           tags$h3("OR"),
           fileInput("coord_file", label = h3("Upload a file with genomic coordinates (BED format)"), 
                     multiple = FALSE, accept = NULL,
                     width = NULL, buttonLabel = "Browse...",
                     placeholder = "No file selected"),
           radioButtons("coord_manual_or_file", "Input type:",
                        c("Manual" = TRUE, "File" = FALSE),
                        selected = FALSE),
           actionButton("coord_submit", label = "Submit"),
           tags$p("Please upload a BED file with one genomic coordinate per line.",
                  "For example:",
                  tags$br(),
                  tags$br(),
                  "chr10\t26216307\t26304562",
                  tags$br(),
                  "chr12\t111034165\t111350554",
                  tags$br(),
                  "chr3\t62369681\t62373550",
                  tags$br(),
                  "chr10\t26216307\t26304562",
                  tags$br(),
                  "..."),
           tags$hr(),
           tags$h3("Advanced Probe Settings (optional)"),
           radioButtons("repeat_seq_coord", "Repeat:",
                        c("allowed" = TRUE, "none" = FALSE),
                        selected = FALSE),
           sliderInput("off_target_coord", "Off-Target Score:",
                       min = 0, max = 1000,
                       value = 200),
           sliderInput("max_kmer_coord", "Max K-mer Count:",
                       min = 0, max = 255,
                       value = 5),
           actionButton("coord_restore_default", label = "Restore Default Parameters"),
           tags$hr(),
           tags$h3("Balance Set (optional)"),
           radioButtons("coord_balance_show", "",
                        c("Show" = TRUE, "Hide" = FALSE),
                        selected = FALSE),
           hidden(sliderInput("coord_balance_goal", "Number of probes per target:",
                       min = 15, max = 1000,
                       value = 100)),
           hidden(radioButtons("coord_balance_set", "Enable:",
                        c("Trim" = 1, "Balance" = 2, "Off" = 3),
                        selected = 3))
         ),
         mainPanel(
           plotOutput("coord_count_plot") %>% withSpinner(color = "grey"),
           DT::dataTableOutput("coord_intersect_table")
           
         )
        )
      ),
      tabPanel("Append Sequences",
        sidebarLayout(
         sidebarPanel(
           # determine which design tab was used
           radioButtons("design_scheme", "Design Scheme:",
                        c("RNA Probe Design" = TRUE, "DNA Probe Design" = FALSE),
                        selected = TRUE),
           tags$hr(),
           ###########################################################################
           radioButtons("fpo_choice", label = h3("5' Outer Primer Sequence (O)"),
                        c("Append" = TRUE, "None" = FALSE),
                        selected = FALSE),
           hidden(radioButtons("fpo_orientation", "Orientation:",
                               c("Forward" = FALSE, 
                                 "Reverse complement" = TRUE),
                               selected = FALSE)),
           hidden(radioButtons("fpo_append_scheme", "Format:",
                        c("Same for all probes" = 1, 
                          "Unique for each target" = 2,
                          "Multiple per target" = 3,
                          "Custom ranges" = 4),
                        selected = 1)),
           hidden(textInput("fpo_custom_ranges", label = "Custom Ranges (optional)", 
                            value = "1-100, 201-300, 301-400, ...")),
           hidden(sliderInput("fpo_n_per_target", "Number Per Target (optional)",
                              min = 1, max = 20, value = 1)),
           hidden(selectInput("fpo_sequence_select", label = "Select Sequence Set", 
                       choices = list("PaintSHOP 5' Outer Primer Set" = 1, "Custom Set" = 2), 
                       selected = 1)),
           hidden(fileInput("fpo_custom_file", label = "Upload Custom Set (optional)", 
                     multiple = FALSE, accept = NULL,
                     width = NULL, buttonLabel = "Browse...",
                     placeholder = "No file selected")),
           tags$hr(),
           ###########################################################################
           radioButtons("fpb_choice", label = h3("5' Bridge Sequence (B)"),
                        c("Append" = TRUE, "None" = FALSE),
                        selected = FALSE),
           hidden(radioButtons("fpb_orientation", "Orientation:",
                               c("Forward" = FALSE, 
                                 "Reverse complement" = TRUE),
                               selected = FALSE)),
           hidden(radioButtons("fpb_append_scheme", "Format:",
                        c("Same for all probes" = 1, 
                          "Unique for each target" = 2,
                          "Multiple per target" = 3,
                          "Custom ranges" = 4),
                        selected = 1)),
           hidden(textInput("fpb_custom_ranges", label = "Custom Ranges (optional)",
                            value = "1-100, 201-300, 301-400, ...")),
           hidden(sliderInput("fpb_n_per_target", "Number Per Target (optional)",
                              min = 1, max = 20, value = 1)),
           hidden(selectInput("fpb_sequence_select", label = "Select Sequence Set", 
                       choices = list("PaintSHOP 5' Bridge Set" = 1, "Custom Set" = 2), 
                       selected = 1)),
           hidden(fileInput("fpb_custom_file", label = "Upload Custom Set (optional)", 
                     multiple = FALSE, accept = NULL,
                     width = NULL, buttonLabel = "Browse...",
                     placeholder = "No file selected")),
           tags$hr(),
           ###########################################################################
           radioButtons("fpi_choice", label = h3("5' Inner Primer Sequence (I)"),
                        c("Append" = TRUE, "None" = FALSE),
                        selected = FALSE),
           hidden(radioButtons("fpi_orientation", "Orientation:",
                               c("Forward" = FALSE, 
                                 "Reverse complement" = TRUE),
                               selected = FALSE)),
           hidden(radioButtons("fpi_append_scheme", "Format:",
                        c("Same for all probes" = 1, 
                          "Unique for each target" = 2,
                          "Multiple per target" = 3,
                          "Custom ranges" = 4),
                        selected = 1)),
           hidden(textInput("fpi_custom_ranges", label = "Custom Ranges (optional)",
                            value = "1-100, 201-300, 301-400, ...")),
           hidden(sliderInput("fpi_n_per_target", "Number Per Target (optional)",
                              min = 1, max = 20, value = 1)),
           hidden(selectInput("fpi_sequence_select", label = "Select Sequence Set", 
                       choices = list("PaintSHOP 5' Primer Set" = 1, "Custom Set" = 2), 
                       selected = 1)),
           hidden(fileInput("fpi_custom_file", label = "Upload Custom Set (optional)", 
                     multiple = FALSE, accept = NULL,
                     width = NULL, buttonLabel = "Browse...",
                     placeholder = "No file selected")),
           tags$hr(),
           ###########################################################################
           
           # determine 3' encoding scheme
           radioButtons("tp_appending_choice", label = h3("3' Appending Scheme"),
                        c("SABER" = 1, "Primer/Bridge/Universal" = 2,
                          "None" = 3),
                        selected = 3),
           tags$hr(),
           ###########################################################################
           
           # SABER Sequence options
           hidden(radioButtons("saber_x", "Number of copies of concatemer:",
                               c("1x" = 1, "2x" = 2),
                               selected = 1)),
           hidden(radioButtons("saber_append_scheme", "Format:",
                               c("Same for all probes" = 1, 
                                 "Unique for each target" = 2,
                                 "Multiple per target" = 3,
                                 "Custom ranges" = 4),
                               selected = 1)),
           hidden(textInput("saber_custom_ranges", label = "Custom Ranges (optional)",
                            value = "1-100, 201-300, 301-400, ...")),
           hidden(sliderInput("saber_n_per_target", "Number Per Target (optional)",
                              min = 1, max = 20, value = 1)),

           ###########################################################################
           hidden(radioButtons("tpi_choice", label = h3("3' Inner Primer Sequence (I)"),
                        c("Append" = TRUE, "None" = FALSE),
                        selected = FALSE)),
           hidden(radioButtons("tpi_orientation", "Orientation:",
                               c("Forward" = FALSE, 
                                 "Reverse complement" = TRUE),
                               selected = TRUE)),
           hidden(radioButtons("tpi_append_scheme", "Format:",
                               c("Same for all probes" = 1, 
                                 "Unique for each target" = 2,
                                 "Multiple per target" = 3,
                                 "Custom ranges" = 4),
                               selected = 1)),
           hidden(textInput("tpi_custom_ranges", label = "Custom Ranges (optional)",
                            value = "1-100, 201-300, 301-400, ...")),
           hidden(sliderInput("tpi_n_per_target", "Number Per Target (optional)",
                              min = 1, max = 20, value = 1)),
           hidden(selectInput("tpi_sequence_select", label = "Select Sequence Set", 
                              choices = list("PaintSHOP 3' Inner Primer Set" = 1, "Custom Set" = 2), 
                              selected = 1)),
           hidden(fileInput("tpi_custom_file", label = "Upload Custom Set (optional)", 
                            multiple = FALSE, accept = NULL,
                            width = NULL, buttonLabel = "Browse...",
                            placeholder = "No file selected")),
           hidden(tags$hr(id = "tpi_choice_hr")),
           ###########################################################################
           hidden(radioButtons("tpb_choice", label = h3("3' Bridge Sequence (B)"),
                        c("Append" = TRUE, "None" = FALSE),
                        selected = FALSE)),
           hidden(radioButtons("tpb_orientation", "Orientation:",
                               c("Forward" = FALSE, 
                                 "Reverse complement" = TRUE),
                               selected = TRUE)),
           hidden(radioButtons("tpb_append_scheme", "Format:",
                               c("Same for all probes" = 1, 
                                 "Unique for each target" = 2,
                                 "Multiple per target" = 3,
                                 "Custom ranges" = 4),
                               selected = 1)),
           hidden(textInput("tpb_custom_ranges", label = "Custom Ranges (optional)",
                            value = "1-100, 201-300, 301-400, ...")),
           hidden(sliderInput("tpb_n_per_target", "Number Per Target (optional)",
                              min = 1, max = 20, value = 1)),
           hidden(selectInput("tpb_sequence_select", label = "Select Sequence Set", 
                              choices = list("PaintSHOP 3' Bridge Set" = 1, "Custom Set" = 2), 
                              selected = 1)),
           hidden(fileInput("tpb_custom_file", label = "Upload Custom Set (optional)", 
                            multiple = FALSE, accept = NULL,
                            width = NULL, buttonLabel = "Browse...",
                            placeholder = "No file selected")),
           hidden(tags$hr(id = "tpb_choice_hr")),
           ###########################################################################
           hidden(radioButtons("tpo_choice", label = h3("3' Outer Primer Sequence (O)"),
                        c("Append" = TRUE, "None" = FALSE),
                        selected = FALSE)),
           hidden(radioButtons("tpo_orientation", "Orientation:",
                               c("Forward" = FALSE, 
                                 "Reverse complement" = TRUE),
                               selected = TRUE)),
           hidden(radioButtons("tpo_append_scheme", "Format:",
                               c("Same for all probes" = 1, 
                                 "Unique for each target" = 2,
                                 "Multiple per target" = 3,
                                 "Custom ranges" = 4),
                               selected = 1)),
           hidden(textInput("tpo_custom_ranges", label = "Custom Ranges (optional)", 
                            value = "1-100, 201-300, 301-400, ...")),
           hidden(sliderInput("tpo_n_per_target", "Number Per Target (optional)",
                              min = 1, max = 20, value = 1)),
           hidden(selectInput("tpo_sequence_select", label = "Select Sequence Set", 
                              choices = list("PaintSHOP 3' Outer Primer Set" = 1, "Custom Set" = 2), 
                              selected = 1)),
           hidden(fileInput("tpo_custom_file", label = "Upload Custom Set (optional)", 
                            multiple = FALSE, accept = NULL,
                            width = NULL, buttonLabel = "Browse...",
                            placeholder = "No file selected")),
           hidden(tags$hr(id = "tpo_choice_hr")),
           ###########################################################################
           actionButton("append_submit", label = "Append")
         ),
         mainPanel(
           # main panel for appending
           tags$img(height = 200,
                    width = 1000,
                    src = "probe-all.png"),
           DT::dataTableOutput("append_table")
         )
        )
      ),
      tabPanel("Download",
        sidebarLayout(
          sidebarPanel(
            radioButtons("appended_tf", "Appended Sequences:",
                         c("Yes" = TRUE, "No" = FALSE),
                         selected = FALSE),
            selectInput("download_choice", "Choose a file to download:",
                        choices = c("None Selected" = 1,
                                    "Order File" = 2,
                                    "Appending File" = 3,
                                    "Full Probe File" = 4),
                        selected = 1),
            downloadButton("download_file", "Download")
          ),
          mainPanel(
            DT::dataTableOutput("download_table")
          )
        )
      )
    )
))
