# PaintSHOP UI logic
# Elliot Hershberg
# July 15, 2019

library(shiny)
library(shinythemes)
library(shinyjs)
library(magrittr)
library(shinycssloaders)
library(shinyBS)

# Define UI for application that draws a histogram
shinyUI(fluidPage(
    
    theme = shinytheme("flatly"),
    useShinyjs(),
    
    # add CSS to the header to make horizontal rules visible
    tags$head(
      tags$style(HTML("hr {border-top: 3px solid #000000}"))
    ),
    
    navbarPage("PaintSHOP",
      tabPanel("About",
        div(tags$img(height = 300,
                 width = 300,
                 src = 'PaintSHOP-logo.png'),
            style="text-align: center;"),
        tags$h1("Welcome to PaintSHOP!"),
        tags$hr(),
        tags$h4("PaintSHOP is an application that enables the interactive design of Oligopaint probe libraries 
               for any experiment. The design process is broken down into a few simple steps:"),
        tags$br(),
        tags$ol(
          tags$li(tags$h4("Quickly retrieve the probes that cover your target(s) of interest")),
          tags$li(tags$h4("Interactively experiment with parameters that determine the specificity 
                          and coverage of your probe set. (optional)")),
          tags$li(tags$h4("Append the additional sequences to your probes that your experiment requires. (optional)")),
          tags$li(tags$h4("Download your newly created order file!"))
        ),
        tags$br(),
        tags$h4("The intended app workflow is shown below. If you want to know more about a specific feature in the app, 
                make sure to hover over it with your mouse to see an additional explanation. Happy FISHing!"),
        tags$br(),
        tags$img(height = 300, width = 1400,
                 src = 'workflow-diagram.png')
      ),
      tabPanel("RNA Probe Design",
        sidebarLayout(
          sidebarPanel(
            selectInput("probeset", "Choose Probe Set:",
                         c("hg38 newBalance" = "probes/refseq/hg38_refseq_newBalance.tsv",
                           "hg19 newBalance" = "probes/refseq/hg19_refseq_newBalance.tsv",
                           "mm10 newBalance" = "probes/refseq/mm10_refseq_newBalance.tsv",
                           "mm9 newBalance" = "probes/refseq/mm9_refseq_newBalance.tsv",
                           "dm6 newBalance" = "probes/refseq/dm6_refseq_newBalance.tsv",
                           "ce11 newBalance" = "probes/refseq/ce11_refseq_newBalance.tsv",
                           "OligoMiner hg38 balance" = "probes/refseq/hg38b_refseq.tsv",
                           "OligoMiner hg19 balance" = "probes/refseq/hg19b_refseq.tsv",
                           "2012 Oligopaints hg19" = "probes/refseq/hg19_2012_refseq.tsv",
                           "iFISH4U 40-mer hg19" = "probes/refseq/iFISH4U_refseq.tsv")),
            textInput("refseq_manual", label = h3("Enter RefSeq IDs manually"), 
                      value = "NM_020309, NM_015267, NM_018008, ..."),
            tags$h3("OR"),
            fileInput("refseq_file", label = h3("Upload a File With RefSeq IDs"), 
                      multiple = FALSE, accept = NULL,
                      width = NULL, buttonLabel = "Browse...",
                      placeholder = "No file selected"),
            radioButtons("refseq_manual_or_file", "Input type:",
                         c("Manual" = TRUE, "File" = FALSE),
                         selected = FALSE),
            actionButton("refseq_submit", label = "Submit"),
            tags$p("Please upload a text file with one RefSeq accession per line.",
                   "For example:",
                   tags$br(),
                   tags$br(),
                   "NM_020309",
                   tags$br(),
                   "NM_015267",
                   tags$br(),
                   "NM_018008",
                   tags$br(),
                   "NM_000818",
                   tags$br(),
                   "..."),
            tags$hr(),
            tags$h3("Advanced Probe Settings (optional)"),
            radioButtons("repeat_seq", "Repeat:",
                         c("allowed" = TRUE, "none" = FALSE),
                         selected = FALSE),
            sliderInput("off_target", "Off-Target Score:",
                        min = 0, max = 1000,
                        value = 200),
            sliderInput("max_kmer", "Max K-mer Count:",
                        min = 0, max = 255,
                        value = 5),
            actionButton("restore_default", label = "Restore Default Parameters"),
            tags$hr(),
            tags$h3("Balance Set (optional)"),
            radioButtons("balance_show", "",
                         c("Show" = TRUE, "Hide" = FALSE),
                         selected = FALSE),
            hidden(sliderInput("balance_goal", "Number of probes per target:",
                        min = 15, max = 60,
                        value = 30)),
            hidden(radioButtons("balance_set", "Enable:",
                         c("Trim" = 1, "Balance" = 2, "Off" = 3),
                         selected = 3))
          ),
          mainPanel(
            plotOutput("count_plot") %>% withSpinner(color = "grey"),
            DT::dataTableOutput("intersect_table")
            
          )
        )
      ),
      tabPanel("DNA Probe Design",
        sidebarLayout(
         sidebarPanel(
           # change button values to path to complete newBalance probe set, not intersect
           selectInput("probeset_coord", "Choose Probe Set:",
                       c("hg38 newBalance" = "probes/all/hg38_all_newBalance.tsv",
                         "hg19 newBalance" = "probes/all/hg19_all_newBalance.tsv",
                         "mm10 newBalance" = "probes/all/mm10_all_newBalance.tsv",
                         "mm9 newBalance" = "probes/all/mm9_all_newBalance.tsv",
                         "dm6 newBalance" = "probes/all/dm6_all_newBalance.tsv",
                         "ce11 newBalance" = "probes/all/ce11_all_newBalance.tsv",
                         "OligoMiner hg38 balance" = "probes/all/hg38b_all.tsv",
                         "OligoMiner hg19 balance" = "probes/all/hg19b_all.tsv",
                         "2012 Oligopaints hg19" = "probes/all/hg19_2012_all.tsv",
                         "iFISH4U 40-mer hg19" = "probes/all/iFISH4U_all.tsv")),
           textInput("coord_manual", label = h3("Enter genomic coordinates manually "), 
                     value = "chr10:26216307-26304562, chr12:111034165-111350554, chr3:62369681-62373550, ..."),
           tags$p("(Please enter chr:start-stop for each target. Optionally, you can specify which strand",
                  " to return, by entering chr:start-stop:+ or chr:start-stop:- for a target.)"),
           tags$h3("OR"),
           fileInput("coord_file", label = h3("Upload a file with genomic coordinates (BED-like format)"), 
                     multiple = FALSE, accept = NULL,
                     width = NULL, buttonLabel = "Browse...",
                     placeholder = "No file selected"),
           radioButtons("coord_manual_or_file", "Input type:",
                        c("Manual" = TRUE, "File" = FALSE),
                        selected = FALSE),
           actionButton("coord_submit", label = "Submit"),
           tags$br(),
           tags$br(),
           tags$p("Please upload a BED-like file with one genomic coordinate per line.",
                  " The strand column is optional.",
                  "For example:", tags$br()),
           pre(includeText("example.bed")),
           tags$hr(),
           tags$h3("Advanced Probe Settings (optional)"),
           radioButtons("repeat_seq_coord", "Repeat:",
                        c("allowed" = TRUE, "none" = FALSE),
                        selected = FALSE),
           sliderInput("off_target_coord", "Off-Target Score:",
                       min = 0, max = 1000,
                       value = 200),
           sliderInput("max_kmer_coord", "Max K-mer Count:",
                       min = 0, max = 255,
                       value = 5),
           actionButton("coord_restore_default", label = "Restore Default Parameters"),
           tags$hr(),
           tags$h3("Balance Set (optional)"),
           radioButtons("coord_balance_show", "",
                        c("Show" = TRUE, "Hide" = FALSE),
                        selected = FALSE),
           hidden(sliderInput("coord_balance_goal", "Number of probes per target:",
                       min = 15, max = 1000,
                       value = 100)),
           hidden(radioButtons("coord_balance_set", "Enable:",
                        c("Trim" = 1, "Balance" = 2, "Off" = 3),
                        selected = 3))
         ),
         mainPanel(
           plotOutput("coord_count_plot") %>% withSpinner(color = "grey"),
           DT::dataTableOutput("coord_intersect_table")
           
         )
        )
      ),
      tabPanel("Append Sequences",
        sidebarLayout(
         sidebarPanel(
           # determine which design tab was used
           radioButtons("design_scheme", "Design Scheme:",
                        c("RNA Probe Design" = TRUE, "DNA Probe Design" = FALSE),
                        selected = TRUE),
           tags$hr(),
           ###########################################################################
           radioButtons("fpo_choice", label = h3("5' Outer Primer Sequence (O)"),
                        c("Append" = TRUE, "None" = FALSE),
                        selected = FALSE),
           hidden(radioButtons("fpo_orientation", "Orientation:",
                               c("Forward" = FALSE, 
                                 "Reverse complement" = TRUE),
                               selected = FALSE)),
           hidden(radioButtons("fpo_append_scheme", "Format:",
                               c("Same for all probes" = 1, 
                                 "Unique for each target" = 2,
                                 "Multiple per target" = 3,
                                 "Custom ranges" = 4),
                               selected = 1)),
           hidden(textInput("fpo_custom_ranges", label = "Custom Ranges (optional)", 
                            value = "1-100, 201-300, 301-400, ...")),
           hidden(sliderInput("fpo_n_per_target", "Number Per Target (optional)",
                              min = 1, max = 20, value = 1)),
           hidden(selectInput("fpo_sequence_select", label = "Select Sequence Set", 
                              choices = list("PaintSHOP 5' Outer Primer Set" = "appending/ps_of.tsv",
                                             "MERFISH Primer Set" = "appending/merfish_primers.tsv",
                                             "Custom Set" = FALSE), 
                              selected = "appending/ps_of.tsv")),
           hidden(fileInput("fpo_custom_file", label = "Upload Custom Set (optional)", 
                            multiple = FALSE, accept = NULL,
                            width = NULL, buttonLabel = "Browse...",
                            placeholder = "No file selected")),
           tags$hr(),
           ###########################################################################
           radioButtons("fpb_choice", label = h3("5' Bridge Sequence (B)"),
                        c("Append" = TRUE, "None" = FALSE),
                        selected = FALSE),
           hidden(radioButtons("fpb_orientation", "Orientation:",
                               c("Forward" = FALSE, 
                                 "Reverse complement" = TRUE),
                               selected = FALSE)),
           hidden(radioButtons("fpb_append_scheme", "Format:",
                               c("Same for all probes" = 1, 
                                 "Unique for each target" = 2,
                                 "Multiple per target" = 3,
                                 "Custom ranges" = 4),
                               selected = 1)),
           hidden(textInput("fpb_custom_ranges", label = "Custom Ranges (optional)",
                            value = "1-100, 201-300, 301-400, ...")),
           hidden(sliderInput("fpb_n_per_target", "Number Per Target (optional)",
                              min = 1, max = 20, value = 1)),
           hidden(selectInput("fpb_sequence_select", label = "Select Sequence Set", 
                              choices = list("PaintSHOP Bridge Set" = "appending/ps_bridges.tsv",
                                             "MERFISH Bridge Set" = "appending/merfish_bridges.tsv",
                                             "Mateo et al. 2019 Bridges" = "appending/Mateo2019_bridges.tsv",
                                             "Xia et al. 2019 Bridges" = "appending/Xia2019_bridges.tsv",
                                             "Custom Set" = FALSE), 
                              selected = "appending/ps_bridges.tsv")),
           hidden(fileInput("fpb_custom_file", label = "Upload Custom Set (optional)", 
                            multiple = FALSE, accept = NULL,
                            width = NULL, buttonLabel = "Browse...",
                            placeholder = "No file selected")),
           tags$hr(),
           ###########################################################################
           radioButtons("fpi_choice", label = h3("5' Inner Primer Sequence (I)"),
                        c("Append" = TRUE, "None" = FALSE),
                        selected = FALSE),
           hidden(radioButtons("fpi_orientation", "Orientation:",
                               c("Forward" = FALSE, 
                                 "Reverse complement" = TRUE),
                               selected = FALSE)),
           hidden(radioButtons("fpi_append_scheme", "Format:",
                               c("Same for all probes" = 1, 
                                 "Unique for each target" = 2,
                                 "Multiple per target" = 3,
                                 "Custom ranges" = 4),
                               selected = 1)),
           hidden(textInput("fpi_custom_ranges", label = "Custom Ranges (optional)",
                            value = "1-100, 201-300, 301-400, ...")),
           hidden(sliderInput("fpi_n_per_target", "Number Per Target (optional)",
                              min = 1, max = 20, value = 1)),
           hidden(selectInput("fpi_sequence_select", label = "Select Sequence Set", 
                              choices = list("PaintSHOP 5' Inner Primer Set" = "appending/ps_if.tsv",
                                             "MERFISH Primer Set" = "appending/merfish_primers.tsv",
                                             "Custom Set" = FALSE), 
                              selected = "appending/ps_if.tsv")),
           hidden(fileInput("fpi_custom_file", label = "Upload Custom Set (optional)", 
                            multiple = FALSE, accept = NULL,
                            width = NULL, buttonLabel = "Browse...",
                            placeholder = "No file selected")),
           tags$hr(),
           ###########################################################################
           
           # determine 3' encoding scheme
           radioButtons("tp_appending_choice", label = h3("3' Appending Scheme"),
                        c("SABER" = 1, "Primer/Bridge/Universal" = 2,
                          "None" = 3),
                        selected = 3),
           tags$hr(),
           ###########################################################################
           
           # SABER Sequence options
           hidden(radioButtons("saber_x", "Number of copies of concatemer:",
                               c("1x" = 1, "2x" = 2),
                               selected = 1)),
           hidden(radioButtons("saber_append_scheme", "Format:",
                               c("Same for all probes" = 1, 
                                 "Unique for each target" = 2,
                                 "Multiple per target" = 3,
                                 "Custom ranges" = 4),
                               selected = 1)),
           hidden(textInput("saber_custom_ranges", label = "Custom Ranges (optional)",
                            value = "1-100, 201-300, 301-400, ...")),
           hidden(sliderInput("saber_n_per_target", "Number Per Target (optional)",
                              min = 1, max = 20, value = 1)),

           ###########################################################################
           hidden(radioButtons("tpi_choice", label = h3("3' Inner Primer Sequence (I)"),
                        c("Append" = TRUE, "None" = FALSE),
                        selected = FALSE)),
           hidden(radioButtons("tpi_orientation", "Orientation:",
                               c("Forward" = FALSE, 
                                 "Reverse complement" = TRUE),
                               selected = TRUE)),
           hidden(radioButtons("tpi_append_scheme", "Format:",
                               c("Same for all probes" = 1, 
                                 "Unique for each target" = 2,
                                 "Multiple per target" = 3,
                                 "Custom ranges" = 4),
                               selected = 1)),
           hidden(textInput("tpi_custom_ranges", label = "Custom Ranges (optional)",
                            value = "1-100, 201-300, 301-400, ...")),
           hidden(sliderInput("tpi_n_per_target", "Number Per Target (optional)",
                              min = 1, max = 20, value = 1)),
           hidden(selectInput("tpi_sequence_select", label = "Select Sequence Set", 
                              choices = list("PaintSHOP 3' Inner Primer Set" = "appending/ps_ir.tsv",
                                             "MERFISH Primer Set" = "appending/merfish_primers.tsv",
                                             "Custom Set" = FALSE), 
                              selected = "appending/ps_ir.tsv")),
           hidden(fileInput("tpi_custom_file", label = "Upload Custom Set (optional)", 
                            multiple = FALSE, accept = NULL,
                            width = NULL, buttonLabel = "Browse...",
                            placeholder = "No file selected")),
           hidden(tags$hr(id = "tpi_choice_hr")),
           ###########################################################################
           hidden(radioButtons("tpb_choice", label = h3("3' Bridge Sequence (B)"),
                        c("Append" = TRUE, "None" = FALSE),
                        selected = FALSE)),
           hidden(radioButtons("tpb_orientation", "Orientation:",
                               c("Forward" = FALSE, 
                                 "Reverse complement" = TRUE),
                               selected = TRUE)),
           hidden(radioButtons("tpb_append_scheme", "Format:",
                               c("Same for all probes" = 1, 
                                 "Unique for each target" = 2,
                                 "Multiple per target" = 3,
                                 "Custom ranges" = 4),
                               selected = 1)),
           hidden(textInput("tpb_custom_ranges", label = "Custom Ranges (optional)",
                            value = "1-100, 201-300, 301-400, ...")),
           hidden(sliderInput("tpb_n_per_target", "Number Per Target (optional)",
                              min = 1, max = 20, value = 1)),
           hidden(selectInput("tpb_sequence_select", label = "Select Sequence Set", 
                              choices = list("PaintSHOP Bridge Set" = "appending/ps_bridges.tsv",
                                             "MERFISH Bridge Set" = "appending/merfish_bridges.tsv",
                                             "Mateo et al. 2019 Bridges" = "appending/Mateo2019_bridges.tsv",
                                             "Xia et al. 2019 Bridges" = "appending/Xia2019_bridges.tsv",
                                             "Custom Set" = FALSE), 
                              selected = "appending/ps_bridges.tsv")),
           hidden(fileInput("tpb_custom_file", label = "Upload Custom Set (optional)", 
                            multiple = FALSE, accept = NULL,
                            width = NULL, buttonLabel = "Browse...",
                            placeholder = "No file selected")),
           hidden(tags$hr(id = "tpb_choice_hr")),
           ###########################################################################
           hidden(radioButtons("tpo_choice", label = h3("3' Outer Primer Sequence (O)"),
                        c("Append" = TRUE, "None" = FALSE),
                        selected = FALSE)),
           hidden(radioButtons("tpo_orientation", "Orientation:",
                               c("Forward" = FALSE, 
                                 "Reverse complement" = TRUE),
                               selected = TRUE)),
           hidden(radioButtons("tpo_append_scheme", "Format:",
                               c("Same for all probes" = 1, 
                                 "Unique for each target" = 2,
                                 "Multiple per target" = 3,
                                 "Custom ranges" = 4),
                               selected = 1)),
           hidden(textInput("tpo_custom_ranges", label = "Custom Ranges (optional)", 
                            value = "1-100, 201-300, 301-400, ...")),
           hidden(sliderInput("tpo_n_per_target", "Number Per Target (optional)",
                              min = 1, max = 20, value = 1)),
           hidden(selectInput("tpo_sequence_select", label = "Select Sequence Set", 
                              choices = list("PaintSHOP 3' Outer Primer Set" = "appending/ps_or.tsv",
                                             "MERFISH Primer Set" = "appending/merfish_primers.tsv",
                                             "Custom Set" = FALSE), 
                              selected = "appending/ps_or.tsv")),
           hidden(fileInput("tpo_custom_file", label = "Upload Custom Set (optional)", 
                            multiple = FALSE, accept = NULL,
                            width = NULL, buttonLabel = "Browse...",
                            placeholder = "No file selected")),
           hidden(tags$hr(id = "tpo_choice_hr")),
           ###########################################################################
           actionButton("append_submit", label = "Append")
         ),
         mainPanel(
           # main panel for appending
           tags$img(height = 200,
                    width = 1000,
                    src = "probe-all.png"),
           DT::dataTableOutput("append_table")
         )
        )
      ),
      tabPanel("Download",
        sidebarLayout(
          sidebarPanel(
            radioButtons("appended_tf", "Appended Sequences:",
                         c("Yes" = TRUE, "No" = FALSE),
                         selected = FALSE),
            selectInput("download_choice", "Choose a file to download:",
                        choices = c("None Selected" = 1,
                                    "Order File" = 2,
                                    "Appending File" = 3,
                                    "Full Probe File" = 4),
                        selected = 1),
            downloadButton("download_file", "Download")
          ),
          mainPanel(
            DT::dataTableOutput("download_table")
          )
        )
      ),
      tabPanel("Resources",
        tags$h1("Download Annotation Probe Sets"),
        selectInput("annotation_probe_choice", "Choose a file to download:",
                   choices = c("hg38 newBalance RefSeq" = "resources/refseq/hg38_refseq_newBalance.zip",
                               "hg19 newBalance RefSeq" = "resources/refseq/hg19_refseq_newBalance.zip",
                               "mm10 newBalance RefSeq" = "resources/refseq/mm10_refseq_newBalance.zip",
                               "mm9 newBalance RefSeq" = "resources/refseq/mm9_refseq_newBalance.zip",
                               "dm6 newBalance RefSeq" = "resources/refseq/dm6_refseq_newBalance.zip",
                               "ce11 newBalance RefSeq" = "resources/refseq/ce11_refseq_newBalance.zip",
                               "OligoMiner hg38 balance RefSeq" = "resources/refseq/hg38b_refseq.zip",
                               "OligoMiner hg19 balance RefSeq" = "resources/refseq/hg19b_refseq.zip",
                               "2012 Oligopaints hg19 RefSeq" = "resources/refseq/hg19_2012_refseq.zip",
                               "iFISH4U 40-mer hg19 RefSeq" = "resources/refseq/iFISH4U_refseq.zip"),
                   selected = "resources/refseq/hg38_refseq_newBalance.zip"),
        downloadButton("annotation_probe_download", "Download"),
        tags$h1("Download Full Probe Sets"),
        selectInput("raw_probe_choice", "Choose a file to download:",
                    choices = c("hg38 newBalance full" = "resources/all/hg38_all_newBalance.zip",
                                "hg19 newBalance full" = "resources/all/hg19_all_newBalance.zip",
                                "mm10 newBalance full" = "resources/all/mm10_all_newBalance.zip",
                                "mm9 newBalance full" = "resources/all/mm9_all_newBalance.zip",
                                "dm6 newBalance full" = "resources/all/dm6_all_newBalance.zip",
                                "ce11 newBalance full" = "resources/all/ce11_all_newBalance.zip",
                                "OligoMiner hg38 balance full" = "resources/all/hg38b_all.zip",
                                "OligoMiner hg19 balance full" = "resources/all/hg19b_all.zip",
                                "2012 Oligopaints hg19 full" = "resources/all/hg19_2012_all.zip",
                                "iFISH4U 40-mer hg19 full" = "resources/all/iFISH4U_all.zip"),
                    selected = "resources/all/hg38_all_newBalance.zip"),
        downloadButton("raw_probe_download", "Download"),
        tags$h1("Download Appending Sequence Sets"),
        selectInput("appending_raw_choice", "Choose a file to download:",
                    choices = c("PaintSHOP 5' Outer Primer Set" = "resources/appending/ps_of.zip",
                                "MERFISH Primer Set" = "resources/appending/merfish_primers.zip",
                                "PaintSHOP Bridge Set" = "resources/appending/ps_bridges.zip",
                                "MERFISH Bridge Set" = "resources/appending/merfish_bridges.zip",
                                "Mateo et al. 2019 Bridges" = "resources/appending/Mateo2019_bridges.zip",
                                "Xia et al. 2019 Bridges" = "resources/appending/Xia2019_bridges.zip",
                                "PaintSHOP 5' Inner Primer Set" = "resources/appending/ps_if.zip",
                                "MERFISH Primer Set" = "resources/appending/merfish_primers.zip",
                                "SABER 1X Sequences" = "resources/appending/saber_1x.zip",
                                "SABER 2X Sequences" = "resources/appending/saber_2x.zip",
                                "PaintSHOP 3' Inner Primer Set" = "resources/appending/ps_ir.zip",
                                "PaintSHOP 3' Outer Primer Set" = "resources/appending/ps_or.zip")),
        downloadButton("appending_raw_download", "Download")
      ),
      tabPanel("Documentation",
        tags$h3("Click on a topic to expand the description", align = "center"),
        tags$br(),
        bsCollapse(id = "tutorial_collapse", multiple = TRUE,
          bsCollapsePanel("RNA Probe Design", 
                         "The RNA Probe Design tab allows you to enter RefSeq annotations ",
                         "in order to retrieve the probes for your target or list of targets. ",
                         "You can either enter the annotations manually into the text box provided, ",
                         "or upload a file with one annotation per line. If entering annotations ",
                         "manually, please separate each entry with a comma and a space, as shown ",
                         "in the example entry. Once you have entered your annotations, select which ",
                         " input option you used, and press the Submit button to retrieve your probes. ",
                         tags$br(),
                         tags$br(),
                         tags$img(width = 500,
                                  height = 150,
                                  src = "annotation-input.png"),
                         tags$br(),
                         tags$br(),
                         "PaintSHOP will return a density plot of probe coverage, and a dynamic table ",
                         "with your probes. The density plot is intended to provide a visual estimate ",
                         "of how well covered your annotations are.",
                         tags$br(),
                         tags$br(),
                         "The table has the following columns: ",
                         tags$hr(),
                         tags$b("refseq:"), 
                         " the RefSeq ID the probe targets",
                         tags$br(),
                         tags$b("chrom:"), 
                         " the chromosome of the probe sequence",
                         tags$br(),
                         tags$b("start:"), 
                         " the start coordinate of the probe sequence",
                         tags$br(),
                         tags$b("stop:"), 
                         " the stop coordinate of the probe sequence",
                         tags$br(),
                         tags$b("sequence:"), 
                         " The DNA sequence of the oligo probe",
                         tags$br(),
                         tags$b("Tm:"), 
                         " The melting temperature of the probe sequence",
                         tags$br(),
                         tags$b("on_target:"), 
                         " The on-target score generated by the Homology Optimization Pipeline",
                         tags$br(),
                         tags$b("off_target:"), 
                         " The off-target score generated by the Homology Optimization Pipeline",
                         tags$br(),
                         tags$b("repeat_seq:"), 
                         " Whether or not the probe sequence contains bases flagged as repetitive by",
                         " RepeatMasker. 0 = False, 1 = True",
                         tags$br(),
                         tags$b("max_kmer:"), 
                         " The maximum number of occurences of all 18-mers in the probe sequence",
                         "in the genome it targets",
                         tags$br(),
                         tags$b("probe_strand:"), 
                         " The strand orientation of the probe sequence. Positive (+) or negative (-)",
                         tags$hr(),
                         tags$b("Note:"), 
                         " If your target is on the + strand, PaintSHOP will automatically return ",
                         "the - strand probe sequence. This is to ensure your FISH experiment works ",
                         "the way you want!",
                         tags$br(),
                         tags$br(),
                         "Also note that the table can be searched, resized, and paged through. ",
                         tags$br(),
                         tags$br(),
                         "For more information about the advanced settings and set balancing features, ",
                         "Please read those topic descriptions.",
                         
                         style = "primary"),
          bsCollapsePanel("DNA Probe Design", 
                         "The RNA Probe Design tab allows you to enter genomic coordinates ",
                         "in order to retrieve the probes for your target or list of targets. ",
                         "You can either enter the coordinates manually into the text box provided, ",
                         "or upload a file with the format shown in the example. If entering coordinates ",
                         "manually, please separate each entry with a comma and a space, as shown ",
                         "in the example entry. Once you have entered your annotations, select which ",
                         " input option you used, and press the Submit button to retrieve your probes. ",
                         tags$br(),
                         tags$br(),
                         tags$img(width = 500,
                                  height = 150,
                                  src = "annotation-input.png"),
                         tags$br(),
                         tags$br(),
                         "PaintSHOP will return a density plot of probe coverage, and a dynamic table ",
                         "with your probes. The density plot is intended to provide a visual estimate ",
                         "of how well covered your targets are.",
                         tags$br(),
                         tags$br(),
                         "The table has the following columns: ",
                         tags$hr(),
                         tags$b("chrom:"), 
                         " the chromosome of the probe sequence",
                         tags$br(),
                         tags$b("start:"), 
                         " the start coordinate of the probe sequence",
                         tags$br(),
                         tags$b("stop:"), 
                         " the stop coordinate of the probe sequence",
                         tags$br(),
                         tags$b("sequence:"), 
                         " The DNA sequence of the oligo probe",
                         tags$br(),
                         tags$b("Tm:"), 
                         " The melting temperature of the probe sequence",
                         tags$br(),
                         tags$b("on_target:"), 
                         " The on-target score generated by the Homology Optimization Pipeline",
                         tags$br(),
                         tags$b("off_target:"), 
                         " The off-target score generated by the Homology Optimization Pipeline",
                         tags$br(),
                         tags$b("repeat_seq:"), 
                         " Whether or not the probe sequence contains bases flagged as repetitive by",
                         " RepeatMasker. 0 = False, 1 = True",
                         tags$br(),
                         tags$b("max_kmer:"), 
                         " The maximum number of occurences of all 18-mers in the probe sequence",
                         "in the genome it targets",
                         tags$br(),
                         tags$b("probe_strand:"), 
                         " The strand orientation of the probe sequence. Positive (+) or negative (-)",
                         tags$hr(),
                         tags$b("Note:"), 
                         " If you specify probe strand (+ or -) in either the manual or file entry, ",
                         "PaintSHOP will return the probe in the same orientation as what you enter. ",
                         "For example specifying + will return +. You are specifying the probe strand, ",
                         "not the target strand.",
                         tags$br(),
                         tags$br(),
                         "Also note that the table can be searched, resized, and paged through. ",
                         tags$br(),
                         tags$br(),
                         "For more information about the advanced settings and set balancing features, ",
                         "Please read those topic descriptions.",
                         
                         style = "primary"),
          bsCollapsePanel("Advanced Probe Settings",
                          "In some cases, you may find that you want to increase the number of probes ",
                          "returned for your targets. PaintSHOP provides a set of advanced features to ",
                          "enable this flexibility. In computational probe design there is an inherent ",
                          "trade-off between probe coverage and specifity. Probe specificity is the ",
                          "likelihood that a probe hybridizes to its intended target instead of at another ",
                          "site in the genome. A greater emphasis on probe specificity inevitably filters ",
                          "probes, reducing coverage. In order to have more control over this design choice, ",
                          "PaintSHOP provides three parameters: 1) repeat inclusion, 2) off-target score, and ",
                          "3) the maximum k-mer count. These three parameters are described below.",
                          tags$hr(),
                          tags$b("Repeat:"),
                          tags$a(href = "http://www.repeatmasker.org/", "RepeatMasker"),
                          " is a program that identifies the presence of repetitive elements in a given DNA ",
                          "sequence. The human genome has been annotated by RepeatMasker. Previous probe design ",
                          "tools have excluded repetitive sequences. PaintSHOP provides the option to allow for ",
                          "probes which contain bases that have been flagged as repetitive, if it is necessary ",
                          "to have enough probes for a target that is challenging to cover. By default, repeat mode ",
                          "is set to off.",
                          tags$br(),
                          tags$br(),
                          tags$b("Off-Target Score:"),
                          " One important component of PaintSHOP is the Homology Optimization Pipeline (HOP). ",
                          "The pipeline is used when creating new probe sets to create an on-target and off-target ",
                          "score for every probe identified. We have developed a machine learning model to approximate ",
                          "nucleic acid thermodynamics. For the on-target score, the model is used to score the ",
                          "likelihood (0-100) that a probe is likely to hybridize at its intended target. ",
                          "For the off-target score, we start by searching for up to 100 possible alignments ",
                          "for each candidate probe. Any candidate with greater than 100 possible alignments is ",
                          "discarded. Next, we use our model to generate a score for the likelihood of hybridization ",
                          "at each possible site. We sum these scores, generating an off-target score between 0 and ",
                          "10,000. By default, PaintSHOP sets the default maximum off-target score to 200. The ",
                          "off-target score slider can be used to make this value more or less stringent, depending ",
                          "on the experiment. The probe table and density plot will dynamically update, providing ",
                          "information on how the parameters chosen affect the probe set.",
                          tags$br(),
                          tags$br(),
                          tags$b("Max K-mer count:"),
                          " A k-mer is a substring of a DNA sequence of length k. We use ",
                          tags$a(href = "https://www.cbcb.umd.edu/software/jellyfish/", "JELLYFISH"),
                          " to count how many times each 18-mer substring in a given probe occurs. ",
                          "in the genome it targets. ",
                          "The maximum value of all 18-mer counts is another way to control probe specifity, ",
                          "And can identify problematic problematic substrings that other alignment approaches ",
                          "may miss. By default, PaintSHOP uses a maximum k-mer count of 5. This can be changed ",
                          " using the slider, and the probe table and density plot will dynamically update.",
                          tags$hr(),
                          tags$b("Note:"),
                          " At any time, the Restore Default Parameters button can be used to reset the default ",
                          "values if you want to remove changes you have made.",
                          
                          style = "primary"),
          bsCollapsePanel("Balancing a Probe Set",
                          "In some instances, you may want to even out your probe set, by selecting a certain ",
                          "number of probes per target. PaintSHOP offers two features for balancing a probe ",
                          "set: 1) trim, and 2) balance. For either option, use the slider to choose how many ",
                          "probes you want per target. The trim option simply ranks the probes for each target ",
                          "by off-target score, and keeps however many you need, removing the rest. The balance ",
                          "option will behave the same way for targets with enough probes, and will automatically ",
                          "relax stringency parameters for targets without enough probes with the currently ",
                          "selected parameters to return enough probes to meet the target, and trim the probes ",
                          "from targets with too many probes.", tags$b("Importantly, "), "this means that ",
                          "changes from the Advanced Probe Settings section will be overriden using the ",
                          "balance feature.",
                          
                          style = "primary"),
          bsCollapsePanel("Appending Sequences",
                          "PaintSHOP provides a suite of functionality for appending the necessary sequences ",
                          "to your probes to carry out your experiment. The following diagram shows where you ",
                          "can append sequences to your probes:",
                          tags$br(),
                          tags$br(),
                          tags$img(height = 200,
                                   width = 800,
                                   src = "probe-all.png"),
                          tags$br(),
                          tags$br(),
                          "The probe sequence itself is the Homology Region (H) in the center. If you are ",
                          "going to amplify your probes from an oligo pool using PCR, you can add forward ",
                          "primers to the 5' I location, and reverse primers to the 3' I location ",
                          "(I stands for Inner Primer). The next regions where sequences can be added ",
                          "are the bridge regions (B) on both the 5' and 3' sides of the probe. ",
                          "Other names for this include read out, ear, and barcode. This sequence can be used ",
                          "as a location for secondary oligos to bind to. Sequences can also be appended to ",
                          "the outer region (O) on the 5' and 3' sides. One useful application of this is ",
                          "to include region/target specific primer pairs to be able to amplify only specific ",
                          "portions of your oligo pool at certain times in your experimental workflow. ",
                          "You can also choose to add ", tags$a(href = "http://saber.fish/", "SABER"),
                          " concatemer sequences to your probes. ",
                          tags$br(),
                          tags$br(),
                          tags$b("Every appending location is optional."),
                          tags$br(),
                          tags$br(),
                          "To append sequences, first select whether you used the RNA Probe Design or DNA ",
                          " Probe Design tab. ",
                          
                          style = "primary")

        )
        
      )
    )
))
